src/test/CharterManager-unit.t.sol

```diff
@@ -0,0 +1,564 @@
+
+        b.join(100 * 1e6);
+        b.frob(100 * 1e18, 50 * 1e18);
+    }
+
+    // Non-msg.sender frobs should be disallowed for now
+    function testFail_frob1() public {
+        (Usr a,) = init_user();
+        a.join(100 * 1e6);
+        a.frob(address(this), address(a), address(a), 100 * 1e18, 50 * 1e18);
+    }
+    function testFail_frob2() public {
+        (Usr a,) = init_user();
+        a.join(100 * 1e6);
+        a.frob(address(a), address(this), address(a), 100 * 1e18, 50 * 1e18);
+    }
+    function testFail_frob3() public {
+        (Usr a,) = init_user();
+        a.join(100 * 1e6);
+        a.frob(address(a), address(a), address(this), 100 * 1e18, 50 * 1e18);
+    }
+
+    function test_frob_other() public {
+        (Usr a, Usr b) = init_user();
+        assertEq(a.gems(), 0);
+        assertEq(b.gems(), 0);
+        assertEq(gem.balanceOf(address(a)), 200 * 1e6);
+        assertEq(gem.balanceOf(address(b)), 200 * 1e6);
+        a.join(address(b), 100 * 1e6);
+        assertEq(gem.balanceOf(address(a)), 100 * 1e6);
+        assertEq(b.gems(), 100 * 1e18);
+        b.allow(address(a));
+        a.frob(address(b), address(b), address(a), 100 * 1e18, 50 * 1e18);
+        assertEq(b.gems(), 0);
+        (uint256 ink, uint256 art) = b.urn();
+        assertEq(ink, 100 * 1e18);
+        assertEq(art, 50 * 1e18);
+        assertEq(a.dai(), 50 * 1e45);
+        assertEq(a.gems(), 0);
+        a.frob(address(b), address(b), address(a), -100 * 1e18, -50 * 1e18);
+        (ink, art) = b.urn();
+        assertEq(ink, 0);
+        assertEq(art, 0);
+        assertEq(a.dai(), 0);
+        assertEq(b.gems(), 100 * 1e18);
+    }
+
+    function testFail_frob_other1() public {
+        (Usr a, Usr b) = init_user();
+        a.join(address(b), 100 * 1e6);
+        a.frob(address(b), address(b), address(a), 100 * 1e18, 50 * 1e18);
+    }
+
+    function testFail_frob_other2() public {
+        (Usr a, Usr b) = init_user();
+        a.join(address(b), 100 * 1e6);
+        b.allow(address(a));
+        b.disallow(address(a));
+        a.frob(address(b), address(b), address(a), 100 * 1e18, 50 * 1e18);
+    }
+
+    function test_flux_to_other() public {
+        (Usr a, Usr b) = init_user();
+        a.join(100 * 1e6);
+        assertEq(a.gems(), 100 * 1e18);
+        a.flux(address(a), address(b), 100 * 1e18);
+        assertEq(b.gems(), 100 * 1e18);
+        b.exit(100 * 1e6);
+        assertEq(b.gems(), 0);
+        assertEq(gem.balanceOf(address(b)), 300 * 1e6);
+    }
+    function test_flux_yourself() public {
+        // Flux to yourself should be a no-op
+        (Usr a,) = init_user();
+        a.join(100 * 1e6);
+        assertEq(a.gems(), 100 * 1e18);
+        a.flux(address(a), address(a), 100 * 1e18);
+        assertEq(a.gems(), 100 * 1e18);
+        a.exit(100 * 1e6);
+        assertEq(a.gems(), 0);
+        assertEq(gem.balanceOf(address(a)), 200 * 1e6);
+    }
+    // Non-msg.sender srcs for flux should be disallowed for now
```

Comment:
This is no longer true as we've added manager delegation via `allow()`.

Reply:
removed that test.